# Step 1: Import Required Libraries

We import the libraries required for data loading,
analysis and visualization.

In [2]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

# Pandas display setting.
pd.set_option('display.max_columns',None)

# Step 2 : Load the Yellow Taxi dataset



In [3]:
df = pd.read_parquet(
    "/home/ed/Desktop/NYC_Taxi_Demand_Forecasting/Yellow_Dataset/data/yellow_tripdata_2026-01.parquet"
)

print("Dataset loaded successfully.")

Dataset loaded successfully.


# Step 3: Dataset Shape

Check the number of rows and columns available in the dataset.

In [4]:
print("Dataset shape:", df.shape)

Dataset shape: (3724889, 20)


# Step 4: Preview Dataset

Display the first few records to understand the dataset structure.

In [5]:
print("First 5 rows of the dataset:")
print(df.head())

First 5 rows of the dataset:
   VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2026-01-01 00:54:04   2026-01-01 00:59:37              1.0   
1         1  2026-01-01 00:34:04   2026-01-01 00:39:47              0.0   
2         1  2026-01-01 00:57:06   2026-01-01 01:05:59              0.0   
3         2  2026-01-01 00:15:22   2026-01-01 00:58:10              4.0   
4         2  2026-01-01 00:27:13   2026-01-01 00:40:43              0.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           0.97         1.0                  N           239           238   
1           0.90         1.0                  N           163           162   
2           1.40         1.0                  N            43           237   
3           5.58         1.0                  N           142           209   
4           2.16         1.0                  N            88           144   

   payment_type  fare_amount  extra  mta_tax 

# Step 5: Dataset Information

Inspect column names, data types, non-null values, and memory usage.

In [6]:
print("Dataset Information::")
df.info()

Dataset Information::
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3724889 entries, 0 to 3724888
Data columns (total 20 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int32         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int32         
 8   DOLocationID           int32         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  

# Step 6: Column Names

Display all available columns in the dataset.

In [7]:
print("Columns in Dataset:")
# print(df.columns.to_list())
for col in df.columns:
    print(col)

Columns in Dataset:
VendorID
tpep_pickup_datetime
tpep_dropoff_datetime
passenger_count
trip_distance
RatecodeID
store_and_fwd_flag
PULocationID
DOLocationID
payment_type
fare_amount
extra
mta_tax
tip_amount
tolls_amount
improvement_surcharge
total_amount
congestion_surcharge
Airport_fee
cbd_congestion_fee


# Step 7: Missing Values

Check whether any columns contain missing values.

In [8]:
missing = df.isnull().sum()
print("Missing values in  column:")
missing[missing > 0].sort_values(ascending=False)

Missing values in  column:


passenger_count         1088058
RatecodeID              1088058
store_and_fwd_flag      1088058
congestion_surcharge    1088058
Airport_fee             1088058
dtype: int64

# Step 8: Missing Value Percentage

Calculate the percentage of missing values in each column.

In [9]:
print("Percentage of missing values in each column:")
(
    df.isnull()
      .mean()
      .mul(100)
      .round(2)
      .sort_values(ascending=False)
)

Percentage of missing values in each column:


passenger_count          29.21
congestion_surcharge     29.21
store_and_fwd_flag       29.21
RatecodeID               29.21
Airport_fee              29.21
tpep_dropoff_datetime     0.00
VendorID                  0.00
tpep_pickup_datetime      0.00
DOLocationID              0.00
payment_type              0.00
trip_distance             0.00
PULocationID              0.00
extra                     0.00
fare_amount               0.00
mta_tax                   0.00
tip_amount                0.00
improvement_surcharge     0.00
tolls_amount              0.00
total_amount              0.00
cbd_congestion_fee        0.00
dtype: float64

# Step 9: Duplicate Records Analysis

Check whether the dataset contains duplicate records.

Duplicate records can lead to incorrect analysis and may affect model performance if not handled properly.


In [10]:
duplicates = df.duplicated().sum()

print(f"Duplicate Rows : {duplicates:,}")

Duplicate Rows : 0


# Step 10: Memory Usage

Measure the memory consumed by the dataset.

This helps estimate the computational resources required for analysis and model training.

In [11]:
memory = (
    df.memory_usage(deep=True)
      .sum()
      /1024 / 1024
)

print(f"Memory Usage : {memory:.2f} MB")

Memory Usage : 647.96 MB


# Step 11: Statistical Summary

Generate summary statistics for numerical columns.

This helps identify data distribution, extreme values, and potential outliers.

In [12]:
print("Descriptive Statistics:")
print(df.describe())

Descriptive Statistics:
           VendorID        tpep_pickup_datetime       tpep_dropoff_datetime  \
count  3.724889e+06                     3724889                     3724889   
mean   1.873598e+00  2026-01-17 01:44:52.106518  2026-01-17 02:02:03.696138   
min    1.000000e+00         2025-12-31 23:57:29         2025-12-31 23:57:32   
25%    2.000000e+00         2026-01-09 17:51:59         2026-01-09 18:08:47   
50%    2.000000e+00         2026-01-16 21:20:56         2026-01-16 21:36:14   
75%    2.000000e+00         2026-01-24 07:25:11         2026-01-24 07:40:22   
max    7.000000e+00         2026-02-01 00:45:01         2026-02-01 23:35:31   
std    7.014583e-01                         NaN                         NaN   

       passenger_count  trip_distance    RatecodeID  PULocationID  \
count     2.636831e+06   3.724889e+06  2.636831e+06  3.724889e+06   
mean      1.256271e+00   6.455647e+00  5.218853e+00  1.614372e+02   
min       0.000000e+00   0.000000e+00  1.000000e+00  1.00

# Step 12: Datetime Range Validation

Check the available date range in the dataset.

This confirms the period covered by the taxi trip records.

In [13]:
print(
    "Minimum Pickup Date:",
    df['tpep_pickup_datetime'].min()
)

print(
    "Maximum Pickup Date:",
    df['tpep_pickup_datetime'].max()
)

Minimum Pickup Date: 2025-12-31 23:57:29
Maximum Pickup Date: 2026-02-01 00:45:01


# Step 13: Forecasting Feasibility Check

Verify that the required columns for hourly demand forecasting are available.

In [14]:
required_cols = [
    'tpep_pickup_datetime',
    'PULocationID'
]

for col in required_cols:
    print(
        f"{col} Exists :",
        col in df.columns
    )

tpep_pickup_datetime Exists : True
PULocationID Exists : True


# Notebook Conclusion

## Key Findings

- Dataset loaded successfully
- Contains approximately 3.7 million trip records
- Data types were verified
- Missing values were identified
- Duplicate records were checked
- Datetime information is available
- Pickup location information is available

## Conclusion

The dataset is suitable for building an hourly taxi demand forecasting model.

## Next Step

Perform Exploratory Data Analysis (EDA) to understand:

- Hourly demand patterns
- Daily demand patterns
- Zone-wise demand distribution
- Feature behavior and trends